### DISCUSSION TITLE
In some cases there seems to be a lot of boilerplate

### DISCUSSION THREAD

**Discussion by dalewking** [badoo/MVIKotlin#202] [General]
While looking into it I find there tends to be a lot of boilerplate. For example, for many intents there is nothing that the executor needs to do and all handling of the intent is just done in the reducer to update the state. So to handle this the executor ends up having to basically just turn the intent into a Result so that the Reducer can update the State. That seems to be a lot of boiler plate. My experience with Redux (in flutter) was much simpler since there wasn't this separation of Intent vs. Action vs. Result. I suppose I could eliminate the distinction myself and just use the same type in for all. Just wondering if you have thought about this

**Comment by dalewking** [comment]
One way I have thought about is that if you want to maintain the type safety of sealed classes for Intent, Action, and Result is make a Result type that is a wrapper for an Intent so for the cases where the only job of the executor is to forward the intent to the reducer it can just wrap the Intent in this wrapper Result.

**Comment by arkivanov** [comment]
If you don't have any async tasks in the `Executor` and if all `Intents` are pure functions, you can just don't add the `Executor` at all. Take a look at the "Simplest example" of the `CalculatorStore` in the [documentation](https://arkivanov.github.io/MVIKotlin/store.html).

However if you need the `Executor` only for some of the Intents, then you can consider the following approach:
```kotlin
class MyStoreFactory(
    private val storeFactory: StoreFactory
) {

    fun create(): MyStore = TODO()

    private sealed class Result {
        data class HandleIntent(val intent: Intent) : Result()
        data class ThisResult(val count: Int) : Result()
        data class ThatResult(val count: Int) : Result()
    }

    private class ExecutorImpl : ReaktiveExecutor<Intent, Nothing, State, Result, Nothing>() {
        override fun executeIntent(intent: Intent, getState: () -> State) =
            when (intent) {
                is Intent.DoAsyncThis -> {
                    // Some logic here
                    dispatch(Result.ThisResult(count = 10))
                }

                is Intent.DoAsyncThat -> {
                    // Some logic here
                    dispatch(Result.ThatResult(count = 20))
                }

                else -> dispatch(Result.HandleIntent(intent))
            }
    }

    private object ReducerImpl : Reducer<State, Result> {
        override fun State.reduce(result: Result): State =
            when (result) {
                is Result.HandleIntent ->
                    when (result.intent) {
                        is Intent.Increment -> copy(count = count + 1)
                        is Intent.Decrement -> copy(count = count - 1)
                        else -> this
                    }

                is Result.ThisResult -> copy(count = result.count)
                is Result.ThatResult -> copy(count = result.count)
            }
    }
}
```

Please not there is less compile time safety here. When you add additional Intents and/or Results, you may forget to cover all the new cases.

**Reply by dalewking** [reply]
What you outline is basically what I described in my second comment about creating a Result to wrap an Intent. I'm still in the exploratory phase of applying MVIKotlin. The one thing I am struggling to get my head around is how to fit the various pieces into dependency injection like Kodein.

**Reply by arkivanov** [reply]
What "various pieces" do you mean? Usually Stores are normal dependencies, working perfectly with DI. You pass things into Stores' constructors, and you pass Stores into higher level things. You can check an example [here](https://github.com/arkivanov/MVIKotlin/blob/master/sample/todo-reaktive/src/commonMain/kotlin/com/arkivanov/mvikotlin/sample/todo/reaktive/controller/TodoListReaktiveController.kt). When using DI frameworks, you should be able to define everything in your DI configuration (e.g. DI modules, etc.), and it will then just assemble everything for you.
I'm ready to help if you highlight any specific problems/issues.

**Reply by dalewking** [reply]
I think I have it sorted out now. The confusion is that StoreFactory is used exactly backwards to how I would usually think of a factory. Normally, I think of a factory as a function that takes no parameters and returns an instance.

I was trying to have my store implementation file create a factory with the details pre-filled, which the DI would call to create the store. That is backwards to the way you have implemented the factory. Instead the implementation details should accept factory from the DI to build the store. So here is what I settled on:

In the file that has the definition of the Executor, Reducer, etc. I expose a function for the DI to call:

```
fun buildFooStore(factory: StoreFactory): FooStore =
    factory.create(
        name = "FooStore",
        initialState = initialState,
        executorFactory = ::FooExecutor
    )
```

The example I took this from is simple and has no reducer or bootstrapper (the executor basically turns intents into Labels).

Then in Kodein I have these bindings:

```
    bind<StoreFactory>() with instance(
        when {
            isDebug -> LoggingStoreFactory(DefaultStoreFactory)
            else -> DefaultStoreFactory
        }
    )

    bind<FooStore>() with factory { buildFooStore(instance()) }
```

**Reply by arkivanov** [reply]
There are two levels of abstraction in MVIKotiln. The first one is the actual implementation of `Store`. There are `DefaultStore`, `LoggingStore` and `TimeTravelStore` provided by the library, and one may create another implementations. This level is managed by `StoreFactory`, which creates an implementation of `Store` given its components: `Bootstrapper`, `Executor` and `Reducer`. The next level is represented by implementations of the `Store`'s components. You implement them and feed to the `StoreFactory`, which creates a `Store` for you. For this level you also need a factory.

So in fact your `buildFooStore` function is a factory function, which creates implementations of the components, feeds them to the `StoreFactory` and returns the result. Usually I create a separate class for my `Stores`:

```kotlin
class MyStoreFactory(
    private val storeFactory: StoreFactory,
    // More dependencies here
) {
    fun create(): MyStore = TODO()

    private inner class Bootstrapper : ...

    private inner class ExcecutorImpl : ...

    private object ReducerImpl : ...
}
```

**Reply by dalewking** [reply]
And I just read the documentation [here](https://arkivanov.github.io/MVIKotlin/store.html) which better showed how to use factory. My bad for not reading the docs

**Reply by dalewking** [reply]
Yes, I switched to a class after reading the documentation:
```
class FooStoreFactory(val factory: StoreFactory, /* additional dependencies */ ): () -> FooStore {
    override fun invoke(): FooStore =
        object: FooStore, Store<Intent,State, Event> by factory.create(
            name = "FooStore",
            initialState = initialState,
            executorFactory = { FooExecutor(/* any additional params */) }
        ) {
        }
}
```

In kodein:

    bind<() -> FooStore>() with singleton { FooStoreFactory(instance()) }

**Reply by arkivanov** [reply]
Yep. Just keep in mind that classes can't extend functions in JavaScript.

**Reply by dalewking** [reply]
Which is not a problem for me as we are just mobile
